libs/angular/src/auth/components/sso.component.ts

```diff
@@ -179,90 +179,157 @@ export class SsoComponent {
+        await this.handleTrustedDeviceEncryptionEnabled(
+          authResult,
+          orgIdentifier,
+          acctDecryptionOpts
+        );
+      } else if (requireSetPassword) {
+        // Change implies going no password -> password in this case
+        await this.handleChangePasswordRequired(orgIdentifier);
+      } else if (authResult.forcePasswordReset !== ForceResetPasswordReason.None) {
+        await this.handleForcePasswordReset(orgIdentifier);
       } else {
-        if (this.onSuccessfulLogin != null) {
-          await this.onSuccessfulLogin();
-        }
-        if (this.onSuccessfulLoginNavigate != null) {
-          await this.onSuccessfulLoginNavigate();
-        } else {
-          this.router.navigate([this.successRoute]);
-        }
+        await this.handleSuccessfulLogin();
       }
     } catch (e) {
-      this.logService.error(e);
-
-      // TODO: Key Connector Service should pass this error message to the logout callback instead of displaying here
-      if (e.message === "Key Connector error") {
-        this.platformUtilsService.showToast(
-          "error",
-          null,
-          this.i18nService.t("ssoKeyConnectorError")
-        );
-      }
+      await this.handleLoginError(e);
     }
     this.loggingIn = false;
   }
 
+  private async isTrustedDeviceEncEnabled(
+    trustedDeviceOption: TrustedDeviceUserDecryptionOption
+  ): Promise<boolean> {
+    const trustedDeviceEncryptionFeatureActive = await this.configService.getFeatureFlagBool(
+      FeatureFlag.TrustedDeviceEncryption
+    );
+
+    return trustedDeviceEncryptionFeatureActive && trustedDeviceOption !== undefined;
+  }
+
+  private async handleTwoFactorRequired(orgIdentifier: string) {
+    await this.navigateViaCallbackOrRoute(
+      this.onSuccessfulLoginTwoFactorNavigate,
+      [this.twoFactorRoute],
+      {
+        queryParams: {
+          identifier: orgIdentifier,
+          sso: "true",
+        },
+      }
+    );
+  }
+
+  private async handleTrustedDeviceEncryptionEnabled(
+    authResult: AuthResult,
+    orgIdentifier: string,
+    acctDecryptionOpts: AccountDecryptionOptions
+  ) {
+    // If user doesn't have a MP, but has reset password permission, they must set a MP
+    if (
+      !acctDecryptionOpts.hasMasterPassword &&
+      acctDecryptionOpts.trustedDeviceOption.hasManageResetPasswordPermission
+    ) {
+      // Change implies going no password -> password in this case
+      await this.handleChangePasswordRequired(orgIdentifier);
+    } else if (authResult.forcePasswordReset !== ForceResetPasswordReason.None) {
+      await this.handleForcePasswordReset(orgIdentifier);
+    } else {
+      // Navigate to TDE page (if user was on trusted device and TDE has decrypted
+      //  their user key, the lock guard will redirect them to the vault)
+      this.router.navigate([this.trustedDeviceEncRoute], {
+        queryParams: {
+          identifier: orgIdentifier,
+        },
+      });
+    }
+  }
```

Comment:
suggestion(non-blocking): write using guard statements

```suggestion
  private async handleTrustedDeviceEncryptionEnabled(
    authResult: AuthResult,
    orgIdentifier: string,
    acctDecryptionOpts: AccountDecryptionOptions
  ): Promise<void> {
    // If user doesn't have a MP, but has reset password permission, they must set a MP
    if (
      !acctDecryptionOpts.hasMasterPassword &&
      acctDecryptionOpts.trustedDeviceOption.hasManageResetPasswordPermission
    ) {
      // Change implies going no password -> password in this case
      return await this.handleChangePasswordRequired(orgIdentifier);
    } 
    
    if (authResult.forcePasswordReset !== ForceResetPasswordReason.None) {
      return await this.handleForcePasswordReset(orgIdentifier);
    } 
    
    // Navigate to TDE page (if user was on trusted device and TDE has decrypted
    //  their user key, the lock guard will redirect them to the vault)
    this.router.navigate([this.trustedDeviceEncRoute], {
      queryParams: {
        identifier: orgIdentifier,
      },
    });
  }
```